shared/src/main/scala/fetch.scala

```diff
@@ -122,21 +118,30 @@ object `package` {
       BlockedRequest(newRequest, newResult)
 
     case (a@Batch(manyId, ds), b@Batch(otherId, _)) =>
-      val newRequest = Batch(combineIdentities(a, b), ds)
+      val combined = combineIdentities(a, b)
+      val newRequest = Batch(combined, ds)
       val newResult = (r: FetchStatus) => (x.result(r), y.result(r)).tupled.void
       BlockedRequest(newRequest, newResult)
   }
 
-  /* A map from datasources to blocked requests used to group requests to the same data source. */
-  private[fetch] final case class RequestMap[F[_]](m: Map[DataSource[Any, Any], BlockedRequest[F]])
+  /* A map from datasource identities to (data source, blocked request) pairs used to group requests to the same data source. */
+  private[fetch] final case class RequestMap[F[_]](
+    m: Map[Data.Identity,
+           (DataSource[F, Any, Any], BlockedRequest[F])])
 
   /* Combine two `RequestMap` instances to batch requests to the same data source. */
   private def combineRequestMaps[F[_] : Monad](x: RequestMap[F], y: RequestMap[F]): RequestMap[F] =
     RequestMap(
       x.m.foldLeft(y.m) {
-        case (acc, (ds, blocked)) => {
-          val combinedReq: BlockedRequest[F] = acc.get(ds).fold(blocked)(combineRequests(blocked, _))
-          acc.updated(ds, combinedReq)
+        case (acc, (dsId, (ds, blocked))) => {
+          val combined = acc.get(dsId).fold(
+            (ds, blocked)
+          )({
+            case (ds, req) => {
```

Comment:
The `ds` variable here shadows the one from line 136. Should they always be the same?

Reply:
Yes.

Reply:
In that case, perhaps we should use back-quotes? As it is right now, the second `ds` is a completely different variable, that just happens to have the same name.